text/3451-additional-float-types.md

```diff
@@ -0,0 +1,139 @@
+let val3: f16 = 1.0;
+let val4 = 1.0f128; // Suffix of f128 literal
+let val5 = 1.0f16; // Suffix of f16 literal
+
+println!("Size of f128 in bytes: {}", std::mem::size_of_val(&val2)); // 16
+println!("Size of f16 in bytes: {}", std::mem::size_of_val(&val3)); // 2
+```
+
+Because not every target supports `f16` and `f128`, compiler provides conditional guards for them:
+
+```rust
+#[cfg(target_has_f128)]
+fn get_f128() -> f128 { 1.0f128 }
+
+#[cfg(target_has_f16)]
+fn get_f16() -> f16 { 1.0f16 }
+```
+
+All operators, constants and math functions defined for `f32` and `f64` in core, are also defined for `f16` and `f128`, and guarded by respective conditional guards.
+
+`f80` type is defined in `core::arch::{x86, x86_64}`. `doubledouble` type is defined in `core::arch::{powerpc, powerpc64}`. `bf16` type is defined in `core::arch::{arm, aarch64, x86, x86_64}`. They do not have literal representation.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## `f16` type
+
+`f16` consists of 1 bit of sign, 5 bits of exponent, 10 bits of mantissa.
+
+The following `From` and `TryFrom` traits are implemented for conversion between `f16` and other types:
+
+```rust
+impl From<f16> for f32 { /* ... */ }
+impl From<f16> for f64 { /* ... */ }
+impl From<bool> for f16 { /* ... */ }
+impl From<u8> for f16 { /* ... */ }
+impl From<i8> for f16 { /* ... */ }
+```
+
+`f16` will generate `half` type in LLVM IR.
+
+## `f128` type
+
+`f128` consists of 1 bit of sign, 15 bits of exponent, 112 bits of mantissa.
+
+`f128` is available for on targets having (1) hardware instructions or software emulation for 128-bit float type; (2) backend support for `f128` type on the target; (3) essential target features enabled (if any).
+
+The list of targets supporting `f128` type may change over time. Initially, it includes `powerpc64le-*`.
+
+The following traits are also implemented for conversion between `f128` and other types:
+
+```rust
+impl From<f16> for f128 { /* ... */ }
+impl From<f32> for f128 { /* ... */ }
+impl From<f64> for f128 { /* ... */ }
+impl From<bool> for f128 { /* ... */ }
+impl From<u8> for f128 { /* ... */ }
+impl From<i8> for f128 { /* ... */ }
+impl From<u16> for f128 { /* ... */ }
+impl From<i16> for f128 { /* ... */ }
+impl From<u32> for f128 { /* ... */ }
+impl From<i32> for f128 { /* ... */ }
+impl From<u64> for f128 { /* ... */ }
+impl From<i64> for f128 { /* ... */ }
+```
+
+`f128` will generate `fp128` type in LLVM IR.
+
+
+`std::simd` defines new vector types with `f16` or `f128` element: `f16x2` `f16x4` `f16x8` `f16x16` `f16x32` `f128x2` `f128x4`.
+
+For `doubledouble` type, conversion intrinsics are available under `core::arch::{powerpc, powerpc64}`. For `f80` type, conversion intrinsics are available under `core::arch::{x86, x86_64}`.
+
+## Architectures specific types
+
+As for non-standard types, `f80` generates `x86_fp80`, `doubledouble` generates `ppc_fp128`, `bf16` generates `bfloat`.
```

Comment:
```suggestion
## Architectures specific types

- `core::arch::{x86, x86_64}::f80` generates LLVM's `x86_fp80`, 80-bit extended precision
- `core::arch::{powerpc, powerpc64}::doubledouble` generates LLVM's `ppc_fp128`, a `f128` emulated
    type via dual `f64`s
- `core::arch::{arm, aarch64, x86, x86_64}bf16` generates LLVM's `bfloat`, 16-bit "brain" floats
    used in AVX and ARMv8.6-A
    
Where possible, `From` will be implemented to convert `f80` and `doubledouble` to `f128`.

## FFI types

`core::ffi::c_longdouble` will always represent whatever `long double` does in C. Rust will defer to the
compiler backend (LLVM) for what exactly this represents, but it will approximately be:

- 80-bit extended precision (f80) on `x86` and `x86_64`: 
- `f64` double precision with MSVC
- `f128` quadruple precision on AArch64
- `doubledouble` on PowerPC
```